### PR TITLE
a first unit test for backward compatibility of Scouting data formats

### PR DESCRIPTION
**Pull request by missirol**
#### PR description:

This PR adds a refined version of the unit test described in https://github.com/cms-sw/cmssw/issues/41040#issuecomment-1470567862 [*]. The aim of the test is to fail whenever non-backward-compatible changes are introduced in the Scouting data formats.

The implementation of the test could certainly be improved (one suggestion is in https://github.com/cms-sw/cmssw/issues/41040#issuecomment-1470689465), but I decided to open a draft PR because I think the test in this PR is better than nothing. (I plan to address https://github.com/cms-sw/cmssw/issues/41040#issuecomment-1470689465, but I don't have an ETA for that.)

[*]

More precisely, there are two unit tests, one for the Run-2 Scouting data formats ("Scouting*"), and one for the Run-3 ones ("Run3Scouting*"). The two tests have the same structure:

 - a 1st step reads O(100) events from an existing EDM file, and writes the Scouting collections to disk (this first step should fail if non-backward-compatible changes are made to the data formats);

 - a 2nd step uses FWLite to print to stdout one entry/object for each of the Scouting data formats being tested. To check that the correct values are read, the test compares the output of step-2 to a reference in the form of a text file. This is currently done for 1 object per data format. It might be necessary to check more than 1 object (and/or more than 1 event), because https://github.com/cms-sw/cmssw/pull/41025#issuecomment-1471777700 showed that some issues can only be spotted when checking multiple objects/events.

#### PR validation:

The new unit tests pass locally.

#### If this PR is a backport, please specify the original PR and why you need to backport that PR. If this PR will be backported, please specify to which release cycle the backport is meant for:

If accepted, it would be good to backport it to at least `CMSSW_13_0_X`.

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

File: DataFormats/Scouting/test/scoutingCollectionsDumper.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+            ('vector<Run3ScoutingMuon>', 'hltScoutingMuonPacker'),
+            ('vector<Run3ScoutingPFJet>', 'hltScoutingPFPacker'),
+            ('vector<Run3ScoutingParticle>', 'hltScoutingPFPacker'),
+            ('vector<Run3ScoutingPhoton>', 'hltScoutingEgammaPacker'),
+            ('vector<Run3ScoutingTrack>', 'hltScoutingTrackPacker'),
+            ('vector<Run3ScoutingVertex>', 'hltScoutingMuonPacker:displacedVtx'),
+            ('vector<Run3ScoutingVertex>', 'hltScoutingPrimaryVertexPacker:primaryVtx')
+        ]
+    else:
+        raise RuntimeError(f'getScoutingProductsList -- invalid key (must be "Scouting", "Run3Scouting"): "{key}"')
```
Comment (missirol) [inline DataFormats/Scouting/test/scoutingCollectionsDumper.py:37]:
```suggestion
        raise RuntimeError(f'getScoutingProductsList -- invalid key (must be "Scouting", or "Run3Scouting"): "{key}"')
```

File: DataFormats/Scouting/test/scoutingCollectionsIO_cfg.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+parser.add_argument('-s', '--nStreams', type = int, help = 'Number of EDM streams',
+                    default = 0)
+
+parser.add_argument('-i', '--inputFiles', nargs = '+', help = 'List of EDM input files',
+                    default = ['/store/mc/Run3Summer22DR/GluGlutoHHto2B2Tau_kl-5p00_kt-1p00_c2-0p00_TuneCP5_13p6TeV_powheg-pythia8/GEN-SIM-RAW/124X_mcRun3_2022_realistic_v12-v2/2550000/bbfb86f3-4073-47e3-967b-059aa6b904ad.root'])
+
+parser.add_argument('-n', '--maxEvents', type = int, help = 'Max number of input events to be processed',
+                    default = 10)
+
+parser.add_argument('--skipEvents', type = int, help = 'Number of input events to skipped',
```
Comment (missirol) [inline DataFormats/Scouting/test/scoutingCollectionsIO_cfg.py:24]:
```suggestion
parser.add_argument('--skipEvents', type = int, help = 'Number of input events to be skipped',
```

**Review by missirol — COMMENTED** [review]
A couple of typos I'll have to fix.

**Comment by missirol** [comment]
please test

(Only fixed a few typos.)

**Comment by missirol** [comment]
I’m wondering how we proceed with this PR. From my point of view, it could be integrated (unfortunately, I haven’t found time yet to work on the ‘EDAnalyzer version’ of the test).

File: DataFormats/Scouting/test/testDataFormatsScoutingRun2.sh [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    cat $3
+    printf "%s\n" "=== End of Log File ==="
+  fi
+  exit $2
+}
+
+# run test job
+TESTDIR="${LOCALTOP}"/src/DataFormats/Scouting/test
+
+cmsRun "${TESTDIR}"/scoutingCollectionsIO_cfg.py -- \
```
Comment (makortel) [inline DataFormats/Scouting/test/testDataFormatsScoutingRun2.sh:17]: These could now be just
```suggestion
cmsRun "${SCRAM_TEST_PATH}"/scoutingCollectionsIO_cfg.py -- \
```

**Comment by makortel** [comment]
> I’m wondering how we proceed with this PR. From my point of view, it could be integrated (unfortunately, I haven’t found time yet to work on the ‘EDAnalyzer version’ of the test).

I suppose these tests are better than nothing, so maybe it would be beneficial to get the tests in, and improve later.

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
Another incarnation of #39803

**Comment by makortel** [comment]
+1

**Comment by perrotta** [comment]
+1

**Pull request merged**